File: python/molscene/scene.py
"""The fluent ``Scene`` facade.

State lives in Rust (``_core.Scene``); this class adds the ergonomic
keyword-argument API and notebook display. Every styling method returns ``self``
so calls chain. Selections are ``ms.select`` values (never strings).
"""

from __future__ import annotations

import json

from . import _core
from ._viewer import DEFAULT_HEIGHT, render_html
from .selection import Selection, select


def _coerce(selection: Selection) -> Selection:
    """Reject string selections with a pointer to the DSL."""
    if isinstance(selection, str):
        raise TypeError(
            "selections are built with ms.select, not strings "
            f"(e.g. ms.select.chain({selection!r}) or ms.select.protein())"
        )
    return selection


# Order of the style tuple exchanged with `_core` (rep_style / set_rep_style).
_STYLE_FIELDS = ("color", "opacity", "scale", "radius", "text")


class Representation:
    """An editable handle on one representation already added to a scene.

    Style fields can be read and reassigned in place after the fact, without
    rebuilding the scene — geometry is recomputed lazily at display time, so
    edits here are reflected by the next ``show`` / ``to_geometry``::

        scene = ms.load("1ubq").cartoon().surface()
        scene.representations[1].opacity = 0.3
        scene.representations[0].color = "spectrum"
        scene.representations[0].selection = ms.select.chain("A")

    ``kind`` is read-only (it picks the geometry path); ``selection`` and the
    style fields ``color`` / ``opacity`` / ``scale`` / ``radius`` / ``text`` are
    read/write. Setting a field to ``None`` clears it back to the default.
    """

    __slots__ = ("_core", "_index")

    def __init__(self, core: _core.Scene, index: int) -> None:
        self._core = core
        self._index = index

    @property
    def kind(self) -> str:
        return self._core.rep_kind(self._index)

    @property
    def selection(self) -> Selection:
        return self._core.rep_selection(self._index)

    @selection.setter
    def selection(self, value: Selection) -> None:
        self._core.set_rep_selection(self._index, _coerce(value))

    def _get(self, field: str):
        style = self._core.rep_style(self._index)
        return style[_STYLE_FIELDS.index(field)]

    def _set(self, field: str, value) -> None:
        style = list(self._core.rep_style(self._index))
        style[_STYLE_FIELDS.index(field)] = value
        self._core.set_rep_style(self._index, *style)

    @property
    def color(self) -> str | None:
        return self._get("color")

    @color.setter
    def color(self, value: str | None) -> None:
        self._set("color", value)

    @property
    def opacity(self) -> float | None:
        return self._get("opacity")

    @opacity.setter
    def opacity(self, value: float | None) -> None:
        self._set("opacity", value)

    @property
    def scale(self) -> float | None:
        return self._get("scale")

    @scale.setter
    def scale(self, value: float | None) -> None:
        self._set("scale", value)

    @property
    def radius(self) -> float | None:
        return self._get("radius")

    @radius.setter
    def radius(self, value: float | None) -> None:
        self._set("radius", value)

    @property
    def text(self) -> str | None:
        return self._get("text")

    @text.setter
    def text(self, value: str | None) -> None:
        self._set("text", value)

    def __repr__(self) -> str:
        return (
            f"<molscene.Representation {self.kind} "
            f"selection={self.selection!s} color={self.color!r} "
            f"opacity={self.opacity!r}>"
        )


class Scene:
    def __init__(self, core: _core.Scene) -> None:
        self._core = core
        self._n_reps = 0

    # -- constructors -------------------------------------------------------
    @classmethod
    def from_rcsb(cls, pdb_id: str, pdb_text: str) -> Scene:
        return cls(_core.Scene.from_rcsb(pdb_id, pdb_text))

    @classmethod
    def from_inline_pdb(cls, pdb_text: str) -> Scene:
        return cls(_core.Scene.from_inline_pdb(pdb_text))

    @classmethod
    def from_inline_sdf(cls, sdf_text: str) -> Scene:
        return cls(_core.Scene.from_inline_sdf(sdf_text))

    # -- representations ----------------------------------------------------
    @property
    def representations(self) -> list[Representation]:
        """Editable handles on the representations added so far, in order.

        Indexable and iterable (``scene.representations[0]``, ``for rep in
        scene.representations``); each item's style fields and selection can be
        reassigned in place — see :class:`Representation`.
        """
        return [
            Representation(self._core, i)
            for i in range(self._core.num_representations())
        ]

    def _add(
        self,
        kind: str,
        selection: Selection,
        *,
        color=None,
        opacity=None,
        scale=None,
        radius=None,
        text=None,
    ) -> Scene:
        self._core.representation(
            kind,
            _coerce(selection),
            color=color,
            opacity=opacity,
            scale=scale,
            radius=radius,
            text=text,
        )
        self._n_reps += 1
        return self

    def cartoon(
        self, selection: Selection | None = None, *, color=None, opacity=None
    ) -> Scene:
        """Protein/nucleic ribbons. Defaults to ``ms.select.protein()`` — crystal
        waters and solvent ions are never drawn as cartoon."""
        sel = selection if selection is not None else select.protein()
        return self._add("cartoon", sel, color=color, opacity=opacity)

    def surface(
        self, selection: Selection | None = None, *, color=None, opacity=None
    ) -> Scene:
        """Molecular surface. Defaults to ``ms.select.protein()`` — water/solvent
        is excluded."""
        sel = selection if selection is not None else select.protein()
        return self._add("surface", sel, color=color, opacity=opacity)

    def sticks(
        self,
        selection: Selection | None = None,
        *,
        color=None,
        opacity=None,
        radius=None,
    ) -> Scene:
        """Bond sticks. Defaults to ``ms.select.ligand()`` (hetero atoms minus
        water), so crystal waters aren't drawn by accident."""
        sel = selection if selection is not None else select.ligand()
        return self._add("sticks", sel, color=color, opacity=opacity, radius=radius)

    def spheres(
        self,
        selection: Selection | None = None,
        *,
        color=None,
        opacity=None,
        scale=None,
    ) -> Scene:
        """Atom spheres. Defaults to everything *except* solvent
        (``ms.select.all() & ~ms.select.solvent()``) so crystal waters and buffer
        ions aren't dumped into the view. Pass ``ms.select.all()`` (or
        ``ms.select.water()`` / ``ms.select.solvent()``) explicitly to show them."""
        sel = selection if selection is not None else select.all() & ~select.solvent()
        return self._add("spheres", sel, color=color, opacity=opacity, scale=scale)

    def lines(
        self,
        selection: Selection | None = None,
        *,
        color=None,
        opacity=None,
        radius=None,
    ) -> Scene:
        """Thin bond lines (a cheap wireframe). Like ``sticks`` but without the
        ball-and-stick atom caps; bond orders are still drawn::

            ms.load("1ubq").lines(ms.select.protein())
        """
        sel = selection if selection is not None else select.all()
        return self._add("lines", sel, color=color, opacity=opacity, radius=radius)

    def dots(
        self,
        selection: Selection | None = None,
        *,
        color=None,
        opacity=None,
        scale=None,
    ) -> Scene:
        """A point cloud: a small sphere per atom (cheaper than ``spheres``)::

        ms.load("1ubq").dots(ms.select.protein())
        """
        sel = selection if selection is not None else select.all()
        return self._add("dots", sel, color=color, opacity=opacity, scale=scale)

    def label(
        self,
        selection: Selection | None = None,
        *,
        text=None,
        color=None,
        size=None,
    ) -> Scene:
        """Text annotations drawn as camera-facing billboards. Defaults to
        ``ms.select.ligand()`` and one label per residue (``"ALA42"``)::

            ms.load("1ubq").cartoon().label(ms.select.resi(50))

        ``text`` picks the content: ``"residue"`` (default), ``"resn"``,
        ``"resi"``, ``"chain"`` (one per residue) or ``"atom"`` / ``"element"``
        (one per atom). Labels default to black; pass ``color`` to recolor and
        ``size`` to scale the font::

            ms.load("lig.sdf").sticks().label(text="atom", size=1.5)
        """
        sel = selection if selection is not None else select.ligand()
        return self._add("labels", sel, text=text, color=color, scale=size)

    def center(self, selection: Selection | None = None) -> Scene:
        """Frame the view on a selection (still auto-fits the zoom)::

        ms.load("1ubq").cartoon().center(ms.select.resi(50))
        """
        if selection is not None:
            self._core.set_center(_coerce(selection))
        return self

    def orient(self, selection: Selection | None = None) -> Scene:
        """Orient the view along a selection's principal axes (PyMOL-style
        ``orient``): its longest dimension goes horizontal, the next vertical::

            ms.load("1ubq").cartoon().orient(ms.select.protein())
        """
        if selection is not None:
            self._core.set_orient(_coerce(selection))
        return self

    def set_color(self, selection: Selection, color: str) -> Scene:
        """Override the color of a sub-selection, on top of the representations'
        schemes. Applied in call order (last write wins), so you can color
        everything one way and then repaint a few atoms::

            ms.load("1ubq").cartoon(color="grey").set_color(ms.select.resi(50), "red")
        """
        self._core.set_color(_coerce(selection), color)
        return self

    def background(self, color: str) -> Scene:
        """Set the scene background color (a named color or ``#rrggbb``).

        Defaults to white when never called::

            ms.load("1ubq").cartoon().background("black")
        """
        self._core.set_background(color)
        return self

    # -- serialization ------------------------------------------------------
    def to_geometry(self) -> dict:
        """The compiled geometry spec (instanced draw list) the renderer draws.

        This is the only serialized form — there is no declarative scene spec;
        the code that builds the scene is the source of truth.
        """
        return json.loads(self._core.to_geometry_json())

    # -- image export -------------------------------------------------------
    def to_png(self, *, width: int = 800, height: int = 600, ssaa: int = 2) -> bytes:
        """Render the scene to PNG bytes with the native Rust GPU rasterizer.

        Unlike :meth:`show` (which hands the geometry to Three.js in the
        browser), this rasterizes the same ``GeometrySpec`` headlessly in Rust
        via ``wgpu`` — spheres and bonds are drawn as ray-traced impostors —
        so it works without a notebook or browser::

            png = ms.load("1ubq").cartoon().to_png(width=1200, height=900)

        ``ssaa`` is the supersampling factor for antialiasing (``1`` disables
        it). Raises ``RuntimeError`` if no GPU (or software fallback such as
        SwiftShader/llvmpipe) is available in the environment.
        """
        return self._core.to_png(width=width, height=height, ssaa=ssaa)

    def save_png(
        self, path: str, *, width: int = 800, height: int = 600, ssaa: int = 2
    ) -> str:
        """Render the scene and write it to ``path`` as a PNG. Returns ``path``::

        ms.load("1ubq").cartoon(color="spectrum").save_png("ubq.png")
        """
        with open(path, "wb") as fh:
            fh.write(self.to_png(width=width, height=height, ssaa=ssaa))
        return path

    # -- display ------------------------------------------------------------
    def _repr_html_(self) -> str:
        return render_html(self.to_geometry())

    def show(self, *, height: int = DEFAULT_HEIGHT, width: str = "100%"):
        from IPython.display import HTML

        return HTML(render_html(self.to_geometry(), height=height, width=width))

    def export_html(
        self, path: str, *, height: int = DEFAULT_HEIGHT, width: str = "100%"
    ) -> str:
        markup = render_html(self.to_geometry(), height=height, width=width)
        with open(path, "w", encoding="utf-8") as fh:
            fh.write(markup)
        return path

    def __repr__(self) -> str:
        return f"<molscene.Scene: {self._n_reps} representation(s)>"

File: crates/molscene-core/src/scene.rs
//! The Scene model: the in-memory model built fluently
//! (`.cartoon().surface().sticks()`) and compiled to a `GeometrySpec` via
//! [`Scene::to_geometry`]. The Scene itself is not serialized — the building code
//! is the source of truth.

use crate::selection::Expr;
use crate::spec::{
    Camera, ColorAssignment, Representation, RepresentationKind, Source, StructureEntry, Style,
};
use crate::structure::Structure;

pub use crate::spec::{Camera as CameraSpec, Representation as RepresentationSpec};
pub use crate::spec::{RepresentationKind as Kind, Source as StructureSource};

/// Id used for the (single, for now) structure.
const STRUCTURE_ID: &str = "s0";

/// A molecular scene: a set of structures, the representations drawn over them,
/// explicit color overrides, and the camera. An in-memory model only.
#[derive(Debug, Clone, PartialEq)]
pub struct Scene {
    structures: Vec<StructureEntry>,
    representations: Vec<Representation>,
    colors: Vec<ColorAssignment>,
    camera: Camera,
    /// Background color, parsed at set-time; `None` keeps the default (white).
    background: Option<crate::color::Rgb>,
    /// Parsed coordinates, kept in memory for native geometry generation.
    structure: Option<Structure>,
}

impl Scene {
    /// Create a scene from a structure source.
    pub fn new(source: Source) -> Self {
        Self {
            structures: vec![StructureEntry {
                id: STRUCTURE_ID.to_string(),
                source,
            }],
            representations: Vec::new(),
            colors: Vec::new(),
            camera: Camera::default(),
            background: None,
            structure: None,
        }
    }

    /// Create a scene that fetches `id` from RCSB (no coordinates loaded yet).
    pub fn from_rcsb(id: impl Into<String>) -> Self {
        Self::new(Source::Rcsb { id: id.into() })
    }

    /// Create a scene from inline PDB text (no coordinates loaded yet).
    pub fn from_inline_pdb(data: impl Into<String>) -> Self {
        Self::new(Source::InlinePdb { data: data.into() })
    }

    /// Parse `text` as an SDF / V2000 molfile and build a scene holding the
    /// resulting coordinates and explicit bond orders.
    #[cfg(feature = "parse")]
    pub fn from_sdf(text: &str, source: Source) -> Result<Self, crate::parse::ParseError> {
        let structure = crate::parse::parse_str(text, crate::parse::InputFormat::Sdf)?;
        let mut scene = Self::new(source);
        scene.structure = Some(structure);
        Ok(scene)
    }

    /// Parse `text` and build a scene that holds the resulting coordinates.
    /// `source` records provenance (e.g. an RCSB id).
    #[cfg(feature = "parse")]
    pub fn from_pdb(text: &str, source: Source) -> Result<Self, crate::parse::ParseError> {
        let structure = crate::parse::parse_str(text, crate::parse::InputFormat::Pdb)?;
        let mut scene = Self::new(source);
        scene.structure = Some(structure);
        Ok(scene)
    }

    /// Attach parsed coordinates to the scene.
    pub fn with_structure(mut self, structure: Structure) -> Self {
        self.structure = Some(structure);
        self
    }

    /// The loaded coordinates, if any.
    pub fn structure(&self) -> Option<&Structure> {
        self.structure.as_ref()
    }

    fn push(&mut self, kind: RepresentationKind, selection: Expr, style: Style) -> &mut Self {
        self.representations.push(Representation {
            structure: STRUCTURE_ID.to_string(),
            kind,
            selection,
            style,
        });
        self
    }

    pub fn cartoon(&mut self, selection: Expr, style: Style) -> &mut Self {
        self.push(RepresentationKind::Cartoon, selection, style)
    }

    pub fn surface(&mut self, selection: Expr, style: Style) -> &mut Self {
        self.push(RepresentationKind::Surface, selection, style)
    }

    pub fn sticks(&mut self, selection: Expr, style: Style) -> &mut Self {
        self.push(RepresentationKind::Sticks, selection, style)
    }

    pub fn spheres(&mut self, selection: Expr, style: Style) -> &mut Self {
        self.push(RepresentationKind::Spheres, selection, style)
    }

    /// Center the camera on a selection (still auto-fits the zoom).
    pub fn center(&mut self, selection: Expr) -> &mut Self {
        self.camera.center = Some(selection);
        self
    }

    /// Orient the view along a selection's principal axes (PyMOL-style
    /// `orient`): its longest dimension is laid out horizontally, the next
    /// vertically. Also frames the selection unless `center` overrides it.
    pub fn orient(&mut self, selection: Expr) -> &mut Self {
        self.camera.orient = Some(selection);
        self
    }

    /// Override the color of a sub-selection, on top of the representations'
    /// schemes. Applied in call order (last write wins) at geometry time.
    pub fn set_color(&mut self, selection: Expr, color: &str) -> &mut Self {
        self.colors.push(ColorAssignment {
            selection,
            color: color.to_string(),
        });
        self
    }

    /// Set the scene background color (a named color or `#rrggbb`). Parsed at
    /// set-time; an unrecognized color leaves the current value unchanged
    /// (the default is white until set).
    pub fn background(&mut self, color: &str) -> &mut Self {
        if let Some(rgb) =
            crate::color::named_color(color).or_else(|| crate::color::parse_hex(color))
        {
            self.background = Some(rgb);
        } else {
            eprintln!("molscene: unknown background color {color:?}; keeping previous value.");
        }
        self
    }

    /// The background color, if one was set.
    pub fn background_color(&self) -> Option<crate::color::Rgb> {
        self.background
    }

    /// Representations added so far.
    pub fn representations(&self) -> &[Representation] {
        &self.representations
    }

    /// Explicit color overrides added so far (in application order).
    pub fn color_assignments(&self) -> &[ColorAssignment] {
        &self.colors
    }

    /// The camera.
    pub fn camera(&self) -> &Camera {
        &self.camera
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn builds_representations_in_order() {
        let mut scene = Scene::from_rcsb("1ubq");
        scene
            .cartoon(
                Expr::Protein,
                Style {
                    color: Some("spectrum".into()),
                    ..Default::default()
                },
            )
            .surface(
                Expr::Protein,
                Style {
                    opacity: Some(0.25),
                    ..Default::default()
                },
            )
            .sticks(
                Expr::Ligand,
                Style {
                    color: Some("element".into()),
                    ..Default::default()
                },
            );

        let reps = scene.representations();
        assert_eq!(reps.len(), 3);
        assert_eq!(reps[0].kind, RepresentationKind::Cartoon);
        assert_eq!(reps[0].selection, Expr::Protein);
        assert_eq!(reps[2].kind, RepresentationKind::Sticks);
        assert_eq!(reps[2].selection, Expr::Ligand);
    }

    #[test]
    fn camera_defaults_to_auto() {
        let scene = Scene::from_rcsb("1ubq");
        assert!(scene.camera().auto);
        assert_eq!(scene.camera().center, None);
        assert_eq!(scene.camera().orient, None);
    }

    #[test]
    fn center_sets_camera_target() {
        let mut scene = Scene::from_rcsb("1ubq");
        scene.center(Expr::Ligand);
        assert_eq!(scene.camera().center, Some(Expr::Ligand));
    }

    #[test]
    fn orient_sets_camera_axes() {
        let mut scene = Scene::from_rcsb("1ubq");
        scene.orient(Expr::Protein);
        assert_eq!(scene.camera().orient, Some(Expr::Protein));
    }

    #[test]
    fn background_parses_named_and_hex_and_keeps_default_on_unknown() {
        let mut scene = Scene::from_rcsb("1ubq");
        assert_eq!(scene.background_color(), None);
        scene.background("black");
        assert_eq!(scene.background_color(), Some([0.0, 0.0, 0.0]));
        scene.background("#ff0000");
        assert_eq!(scene.background_color(), Some([1.0, 0.0, 0.0]));
        // An unrecognized color leaves the previous value untouched.
        scene.background("definitely-not-a-color");
        assert_eq!(scene.background_color(), Some([1.0, 0.0, 0.0]));
    }

    #[test]
    fn set_color_records_overrides_in_order() {
        let mut scene = Scene::from_rcsb("1ubq");
        scene
            .set_color(Expr::Protein, "grey")
            .set_color(Expr::resi(50, 50), "red");
        let overrides = scene.color_assignments();
        assert_eq!(overrides.len(), 2);
        assert_eq!(overrides[0].selection, Expr::Protein);
        assert_eq!(overrides[0].color, "grey");
        assert_eq!(overrides[1].color, "red");
    }
}
